data_model/src/block_value.rs

```diff
@@ -0,0 +1,82 @@
+//! This module contains [`BlockValue`] and [`BlockHeaderValue`] structures, their implementation and related traits and
+//! instructions implementations.
+#[cfg(not(feature = "std"))]
+use alloc::Vec;
+use core::cmp::Ordering;
+
+use iroha_crypto::{Hash, HashOf, MerkleTree};
+use iroha_schema::IntoSchema;
+use parity_scale_codec::{Decode, Encode};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    events::Event,
+    transaction::{TransactionValue, VersionedTransaction},
+};
+
+/// Block header
+#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+pub struct BlockHeaderValue {
+    /// Unix time (in milliseconds) of block forming by a peer.
+    pub timestamp: u128,
+    /// a number of blocks in the chain up to the block.
+    pub height: u64,
+    /// Hash of a previous block in the chain.
+    /// Is an array of zeros for the first block.
+    pub previous_block_hash: Hash,
+    /// Hash of merkle tree root of the tree of valid transactions' hashes.
+    pub transactions_hash: HashOf<MerkleTree<VersionedTransaction>>,
+    /// Hash of merkle tree root of the tree of rejected transactions' hashes.
+    pub rejected_transactions_hash: HashOf<MerkleTree<VersionedTransaction>>,
+    /// Hashes of the blocks that were rejected by consensus.
+    pub invalidated_blocks_hashes: Vec<Hash>,
+}
+
+impl PartialOrd for BlockHeaderValue {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.timestamp.cmp(&other.timestamp))
+    }
+}
+
+impl Ord for BlockHeaderValue {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.timestamp.cmp(&other.timestamp)
+    }
+}
+
+/// Representation of block on blockchain
+#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Serialize, Deserialize, IntoSchema)]
+pub struct BlockValue {
+    /// Header
+    pub header: BlockHeaderValue,
+    /// Array of transactions
+    pub transactions: Vec<TransactionValue>,
+    /// Event recommendations
+    pub event_recommendations: Vec<Event>,
+}
+
+impl BlockValue {
+    /// ...
+    pub fn nested_len(&self) -> usize {
```

Comment:
I'm not sure I fully understand the semantics here, so I'm just adding lengths of everything having a length here :)